Main.hs

```diff
@@ -273,32 +288,86 @@ compileConstructorArgs (ExtendTel a tel)
       x = hsName $ prettyShow n
   setCurrentRange (nameBindingSite n) $ do
     ty <- compileType (unEl $ defType def)
-    cs <- mapM (compileClause x) funClauses
+    let body = hsError $ "postulate: " ++ pp ty
+    return [ Hs.TypeSig () [x] ty
+           , Hs.FunBind () [Hs.Match () x [] (Hs.UnGuardedRhs () body) Nothing] ]
+  where
+    Axiom = theDef def
+
+type LocalDecls = [(QName, Definition)]
+
+compileFun :: Definition -> TCM [Hs.Decl ()]
+compileFun d = do
+  locals <- takeWhile (isAnonymousModuleName . qnameModule . fst)
+          . dropWhile ((<= defName d) . fst)
+          . sortDefs <$> curDefs
+  compileFun' d locals
+
+compileFun' :: Definition -> LocalDecls -> TCM [Hs.Decl ()]
+compileFun' (Defn {..}) locals = do
+  let n = qnameName defName
+      x = hsName $ prettyShow n
+      go = foldM $ \(ds, ms) -> compileClause ds defName x >=> return . fmap (ms `snoc`)
+  setCurrentRange (nameBindingSite n) $ do
+    ty <- compileType (unEl defType)
+    cs <- snd <$> go (locals, []) funClauses
     return [Hs.TypeSig () [x] ty, Hs.FunBind () cs]
   where
-    Function{..} = theDef def
+    Function{..} = theDef
 
-compileClause :: Hs.Name () -> Clause -> TCM (Hs.Match ())
-compileClause x Clause{clauseTel       = tel,
-                       namedClausePats = ps,
-                       clauseBody      = body } =
+compileClause :: LocalDecls -> QName -> Hs.Name () -> Clause -> TCM (LocalDecls, Hs.Match ())
+compileClause locals qn x c@Clause{clauseTel = tel, namedClausePats = ps', clauseBody = body} =
   addContext (KeepNames tel) $ localScope $ do
-    -- Only bind user-written variables
+    -- Compile patterns, only bind user-written variables
     let bind (d, i) | getOrigin d == UserWritten = bindVar i
                     | otherwise                  = return ()
     mapM_ bind $ zip (telToList tel) $ reverse [0..size tel - 1]
-    ps   <- compilePats ps
-    body <- maybe (return $ Hs.Var () $ Hs.UnQual () (hsName "undefined")) compileTerm body
-    let rhs = Hs.UnGuardedRhs () body
-    case (x, ps) of
-      (Hs.Symbol{}, p : q : ps) -> return $ Hs.InfixMatch () p x (q : ps) rhs Nothing
-      _                         -> return $ Hs.Match () x ps rhs Nothing
+    ps <- compilePats ps'
+
+    -- Compile rhs and its @where@ clauses, making sure that:
+    --   * inner modules get instantiated
+    --   * references to inner modules get un-qualifiyed (and instantiated)
+    let localUses = nub $ listify (`elem` map fst locals) body
+        belongs q@(QName m _) (QName m0 _) =
+          ((show m0 ++ "._") `isPrefixOf` show m) && (q `notElem` localUses)
+        splitDecls :: LocalDecls -> ([(Definition, LocalDecls)], LocalDecls)
+        splitDecls ds@((q,child):rest)
+          | q `elem` localUses
+          , (grandchildren, outer) <- span ((`belongs` q) . fst) rest
+          , (groups, rest') <- splitDecls outer
+          = ((child, grandchildren) : groups, rest')
+          | otherwise = ([], ds)
```

Comment:
This is too simplistic. If the first local decl is not in `localUses` it stops. Example:
```agda
ex8 : Nat
ex8 = n₂
  where
    n₁ = 1
    n₂ = n₁ + 1
```
compiles to
```haskell
ex8 :: Integer
ex8 = Where.n
```

Reply:
Changing `elem` to `any elem` did the job.